src/components/Form.js

```diff
@@ -0,0 +1,216 @@
+        const values = this.getValues();
+        if (!_.isEmpty(this.validate(values))) {
+            return;
+        }
+
+        // Set loading state and call submit handler
+        FormActions.setIsSubmitting(this.props.formID, true);
+        this.props.onSubmit(values);
+    }
+
+    /**
+     * @param {Object} values - An object containing the value of each inputID, e.g. {inputID1: value1, inputID2: value2}
+     * @returns {Object} - An object containing the errors for each inputID, e.g. {inputID1: error1, inputID2: error2}
+     */
+    validate(values) {
+        FormActions.setServerErrorMessage(this.props.formID, '');
+        const validationErrors = this.props.validate(values);
+
+        if (!_.isObject(validationErrors)) {
+            throw new Error('Validate callback must return an empty object or an object with shape {inputID: error}');
+        }
+
+        const errors = _.pick(validationErrors, (inputValue, inputID) => (
+            Boolean(this.touchedInputs[inputID])
+        ));
+        this.setState({errors});
+        return errors;
+    }
+
+    childrenWrapperWithProps(children) {
+        return React.Children.map(children, (child) => {
+            // Just render the child if it is not a valid React element, e.g. text within a <Text> component
+            if (!React.isValidElement(child)) {
+                return child;
+            }
+
+            // Depth first traversal of the render tree as the input element is likely to be the last node
+            if (child.props.children) {
+                return React.cloneElement(child, {
+                    children: this.childrenWrapperWithProps(child.props.children),
+                });
+            }
+
+            // We check if the child has the isFormInput prop.
+            // We don't want to pass form props to non form components, e.g. View, Text, etc
+            if (!child.props.isFormInput) {
+                return child;
+            }
+
+            // We clone the child passing down all form props
+            const inputID = child.props.inputID;
+            return React.cloneElement(child, {
+                ref: node => this.inputRefs[inputID] = node,
+                defaultValue: this.props.draftValues[inputID] || child.props.defaultValue,
+                errorText: this.state.errors[inputID] || '',
+                onBlur: (event) => {
+                    this.setTouchedInput(inputID);
+                    this.validate(this.getValues());
+                    if (child.props.onBlur) {
+                        child.props.onBlur(event);
+                    }
+                },
+                onChange: (value) => {
+                    if (child.props.shouldSaveDraft) {
+                        FormActions.setDraftValues(this.props.formID, {[inputID]: value});
+                    }
+                    if (this.touchedInputs[inputID]) {
+                        this.validate(this.getValues());
+                    }
```

Comment:
NAB, Why prevent validating here if the input has not been touched? We already look to see if the inputs have been touched or not in `this.validate()`? Seems like a pre-optimization and it would be better to just leave this up to `this.validate()` (we already do it when gathering the errors https://github.com/Expensify/App/pull/6412/files#diff-a54e851a201173befe09914aee40247c3825f28fa4ff3f01eeaa500be75c5dc8R123-R125)